src/vhost_kern/vdpa.rs

```diff
@@ -0,0 +1,375 @@
+// Copyright (C) 2021 Red Hat, Inc. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0 or BSD-3-Clause
+
+//! Kernel-based vhost-vdpa backend.
+
+use std::fs::{File, OpenOptions};
+use std::os::raw::c_int;
+use std::os::unix::fs::OpenOptionsExt;
+use std::os::unix::io::{AsRawFd, RawFd};
+
+use vm_memory::GuestAddressSpace;
+use vmm_sys_util::eventfd::EventFd;
+use vmm_sys_util::ioctl::{ioctl_with_mut_ref, ioctl_with_ptr, ioctl_with_ref};
+
+use std::alloc::{alloc, dealloc, Layout};
+use std::mem;
+
+use super::vhost_binding::*;
+use super::{ioctl_result, Error, Result, VhostKernBackend, VhostKernFeatures};
+use crate::vdpa::*;
+use crate::{VhostAccess, VhostIotlbBackend, VhostIotlbMsg, VhostIotlbType};
+
+/// Handle for running VHOST_VDPA ioctls.
+pub struct VhostKernVdpa<AS: GuestAddressSpace> {
+    fd: File,
+    mem: AS,
+    backend_features_acked: u64,
+}
+
+impl<AS: GuestAddressSpace> VhostKernVdpa<AS> {
+    /// Open a handle to a new VHOST-VDPA instance.
+    pub fn new(path: &str, mem: AS) -> Result<Self> {
+        Ok(VhostKernVdpa {
+            fd: OpenOptions::new()
+                .read(true)
+                .write(true)
+                .custom_flags(libc::O_CLOEXEC | libc::O_NONBLOCK)
+                .open(path)
+                .map_err(Error::VhostOpen)?,
+            mem,
+            backend_features_acked: 0,
+        })
+    }
+}
+
+impl<AS: GuestAddressSpace> VhostVdpa for VhostKernVdpa<AS> {
+    fn get_device_id(&self) -> Result<u32> {
+        let mut device_id: u32 = 0;
+        let ret = unsafe { ioctl_with_mut_ref(self, VHOST_VDPA_GET_DEVICE_ID(), &mut device_id) };
+        ioctl_result(ret, device_id)
+    }
+
+    fn get_status(&self) -> Result<u8> {
+        let mut status: u8 = 0;
+        let ret = unsafe { ioctl_with_mut_ref(self, VHOST_VDPA_GET_STATUS(), &mut status) };
+        ioctl_result(ret, status)
+    }
+
+    fn set_status(&self, status: u8) -> Result<()> {
+        let ret = unsafe { ioctl_with_ref(self, VHOST_VDPA_SET_STATUS(), &status) };
+        ioctl_result(ret, ())
+    }
+
+    fn get_config(&self, offset: u32, buffer: &mut [u8]) -> Result<()> {
+        let buffer_len = buffer.len();
+        let layout =
+            Layout::from_size_align(mem::size_of::<vhost_vdpa_config>() + buffer_len, 1).unwrap();
```

Comment:
should we use vmm_sys_util::fam_struct  here?

Reply:
Oh absolutely, I didn't know it. I'll give it a try,